File: 005-linked-list/single-linked-list/single-linked-list.go
package main

import (
	"fmt"
)

type node struct {
	key  int
	next *node
}

type singleLinkedList struct {
	head *node
	tail *node
}

func (l *singleLinkedList) insertAtHead(key int) {
	node := node{
		key: key,
	}
	if l.head == nil {
		l.head = &node
		l.tail = &node
	} else {
		node.next = l.head
		l.head = &node
	}
}

func (l *singleLinkedList) insertAtTail(key int) {
	node := node{
		key: key,
	}
	if l.tail == nil {
		l.tail = &node
		l.head = &node
	} else {
		l.tail.next = &node
		l.tail = &node
	}
}

func (l *singleLinkedList) search(key int) (*node, error) {
	ptr := l.head
	for ptr != nil && ptr.key != key {
		ptr = ptr.next
	}
	if ptr == nil {
		return nil, fmt.Errorf("key not found: %d", key)
	}
	return ptr, nil
}

func (l *singleLinkedList) traverse() []int {
	ptr := l.head
	result := []int{}
	for ptr != nil {
		result = append(result, ptr.key)
		ptr = ptr.next
	}
	return result
}

func (l *singleLinkedList) delete(key int) (int, error) {
	var ptr1 *node = nil
	ptr2 := l.head
	for ptr2 != nil && ptr2.key != key {
		ptr1 = ptr2
		ptr2 = ptr2.next
	}
	if ptr2 != nil {
		if ptr1 == nil {
			l.head = ptr2.next
		} else {
			ptr1.next = ptr2.next
		}

		if l.tail == ptr2 {
			l.tail = ptr1
		}
		return key, nil
	}
	return -1, fmt.Errorf("key not found so can not be deleted: %d", key)
}

func new() singleLinkedList {
	l := singleLinkedList{}
	return l
}

func main() {
	l := new()

	fmt.Println("list", l.traverse())

	_, err3 := l.delete(11)
	if err3 != nil {
		fmt.Println(err3)
	}

	l.insertAtHead(4)
	v3, err3 := l.delete(4)
	if err3 != nil {
		fmt.Println(err3)
	} else {
		fmt.Println("key deleted", v3)
		fmt.Println("list after deleting", v3, l.traverse())
	}
	l.insertAtTail(8)
	l.insertAtTail(5)
	l.insertAtHead(3)
	l.insertAtHead(7)

	fmt.Println("list", l.traverse())

	_, err := l.search(6)
	if err != nil {
		fmt.Println(err)
	}

	v, _ := l.search(3)
	fmt.Println("key found", v.key)

	v2, _ := l.delete(5)
	fmt.Println("Key deleted", v2)
	fmt.Println("list after deleting 5 ", l.traverse())

	l.insertAtTail(12)
	fmt.Println("list after inserting 12 at tail ", l.traverse())

	_, err2 := l.delete(18)
	if err2 != nil {
		fmt.Println(err2)
	}
}

File: 005-linked-list/single-linked-list/single-linked-list.js
class Node {
  constructor(key) {
    this.key = key;
    this.next = null;
  }
}

class SingleLinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
  }

  insertAtHead(key) {
    const node = new Node(key);
    if (this.head == null) {
      this.head = node;
      this.tail = node;
    } else {
      node.next = this.head;
      this.head = node;
    }
  }

  insertAtTail(key) {
    const node = new Node(key);
    if (this.tail == null) {
      this.head = node;
      this.tail = node;
    } else {
      this.tail.next = node;
      this.tail = node;
    }
  }

  search(key) {
    let ptr = this.head;
    while (ptr != null && ptr.key != key) {
      ptr = ptr.next;
    }
    if (ptr == null) {
      throw new Error("Key not found");
    } else {
      return ptr;
    }
  }

  delete(key) {
    let ptr1 = null;
    let ptr2 = this.head;
    while (ptr2 != null && ptr2.key != key) {
      ptr1 = ptr2;
      ptr2 = ptr2.next;
    }
    if (ptr2 != null) {
      if (ptr1 == null) {
        this.head = ptr2.next;
      } else {
        ptr1.next = ptr2.next;
      }
      if (this.tail == ptr2) {
        this.tail = ptr1;
      }
      return "Key deleted"
    } else {
      throw new Error("Key not found, so can not be deleted");
    }
  }

  traverse() {
    const result = []
    let ptr = this.head
    while(ptr != null) {
      result.push(ptr.key)
      ptr = ptr.next
    }
    return result
  }
}

const list = new SingleLinkedList()

console.log("list : ", list.traverse())

list.insertAtHead(2)
list.insertAtHead(7)
list.insertAtTail(6)
list.insertAtHead(3)
list.insertAtTail(9)

console.log("list : ", list.traverse())

try {
  console.log("search 3 ", list.search(3))
} catch(e) {
  console.log("search 3 error", e)
}


try {
  console.log("search 17 ", list.search(17))
} catch(e) {
  console.log("search 17 error", e)
}

console.log("delete 3 ", list.delete(3))
console.log("list after delete 3 : ", list.traverse())

console.log("delete 7 ", list.delete(7))
console.log("list after delete 7 : ", list.traverse())

console.log("delete 9 ", list.delete(9))
console.log("list after delete 9 : ", list.traverse())

list.insertAtTail(11)
console.log("list after insert at tail 11 : ", list.traverse())

try {
  console.log("delete 17 ", list.delete(17))
  console.log("list after delete 17 : ", list.traverse())
} catch(e) {
  console.log("delete 17 error ", e)
  console.log("list after delete 17 error : ", list.traverse())
}

